### ISSUE TITLE
Design of seeds with more than 32 bits of randomness

### ISSUE THREAD

**Issue by LTLA** [daqana/dqrng#11]
As suggested by @rstub on #10, I'll try to justify why I chose raw vectors for generating >32 bit seeds.

# Why raw vectors?

Raw vectors are `unsigned char` vectors represented in R, with size in bytes equal to its length (plus some overhead for R vectors). This provides a nice representation in which one can specify the settings for an arbitrary number of bits. Each entry of the raw vector stores 8 bits, so to store more bits, one just uses a longer vector. This is easy to construct in R (see the code in #10); it is also easy to reason with in C++ requiring only a few bit shifts and masks to create a `uintX_t` seed integer from a raw vector.

Let's now consider the integer as an alternative. R uses a 32-bit signed integer representation, so we would need to use two integers to generate a 64-bit seed. Thus, we could use an integer vector instead of a raw vector. Generation of this seed in R could then feasibly be done with:

```r
sample((-.Machine$integer.max):(.Machine$integer.max), 2, replace=TRUE)
```

... which would then be converted into unsigned integers and, with some bit operations, used to generate a 64-bit seed. The problem is that the above code will not span the entire 32-bit range of possible signed integers, as `-INT_MAX` is lost to represent `NA_INTEGER` in R. The correct way to do it:

```r
X <- c(NA, (-.Machine$integer.max):(.Machine$integer.max))
sample(X, 2, replace=TRUE)
```

... causes a `Error: cannot allocate vector of size 16.0 Gb`. And, to be honest, I find reasoning about signed-to-unsigned integer conversions to be difficult, though this might just be my problem.

The final numerical alternative is to use a `double`. My first attempt was to actually generate a number within the range of possible 64-bit unsigned integers, and pass that to C++:

```
runif(1, 0, 2^64-1)
```

The problem here is that I am not guaranteed to sample from the full space of seeds, as double-precision numbers cannot represent every 64-bit integer perfectly. There will be some integers that simply will never be sampled, i.e., `runif` cannot generate numbers within `[i, i+1)` for some integers `i`. The coercion from `double` to `uintX_t` in C++ is also difficult as we need to check that the number can fit within the unsigned range; this is not straightforward if we cannot accurately represent the maximum unsigned number as a double for comparison. (And the penalty of not doing so is extreme - undefined behaviour!)

My second, much shorter-lived attempt was to try to directly convert the bits in a floating point number to a 64-bit unsigned integer. This is probably not a good idea - IIRC, it gets into implementation-specific details about how data are represented on different platforms. Moreover, even if this were portably possible, it is not clear how to sample a floating point number in a way that would get an equal chance of setting each bit.

So, by process of elimination, I ended up with a design using raw vectors. I have not considered using strings because `\0` would need to be sampled to offer 8 bits of randomness per character.

# Why is `generateRawSeeds()` an R function?

This is motivated by the usage of PRNGs in R packages. As a R package developer, I use _dqrng_'s PRNGs in my C++ code, seeded to ensure reproducible execution. However, I **also** want my functions to respond to the state of R's seed, so that users running my functions multiple times (e.g., to check the variance due to the calculations) will get different results. A natural solution to both of these constraints is to randomly define seeds in R and then pass those seeds to the C++ code for seeding and further random number generation.

In this context, it is sensible to have `generateRawSeeds()` as an R function. I can use this function in my package functions to generate seeds in a manner that responds to user-specified `set.seed()`. As an added bonus, it also allows users in interactive sessions to call `generateRawSeeds()` to seed _dqrng_'s R-visible distribution functions with 64-bit of randomness. (Possibly want to set `bits=64` by default).

It is possible to translate this into a C++ function, and just have the R function as a wrapper. This would allow generation of raw-vector seeds from C++ as well, which might be desirable in some settings. I have not done so because... well, I guess I didn't really need it. A direct translation will also have some issues with replacing `sample` with `std::shuffle` due to the latter's lack of reproducibility across platforms. ~One _could_ use `boost::random::uniform_int_distribution` if R's API exposed the raw PRNG... which it doesn't seem to. In _scran_, I have added my own C++ shuffling algorithm to ensure that I get the same result across platforms, but this was more of a change than I was willing to make here without further approval.~ It should be possible to do something fairly simple with `R::unif_rand() * 256` if this is considered to be necessary.

**Comment by LTLA** [comment]
Thanks for the comments @rstub.

# Using raw vectors as the seed 

I can empathize with your wariness towards raw vectors, as I felt the same way when I first saw them. (What's this `0e f6 a3` nonsense? Must be a segmentation fault!) But I have grudgingly come around to concede that they are useful in some cases. I'll start by pointing out that you can directly create raw-vector seeds:

```r
dqset.seed(as.raw(c(0x10, 0x20, 0x30, 0x40)))
```

This is about as interpretable as a integer-vector seed; the values aren't very user-friendly in either case. Yes, it's more typing, but someone going for an easy manual initialization should just supply an integer scalar directly; they'll never do it enough times to run into statistical biases from the limited state space.

If manual usability is critical, integer vectors are probably better than the `bit64` approach. The latter requires an additional `Depends` and doesn't allow us to avoid using vectors (e.g., if you're seeding one of the PRNGs with 128 bits of state in your package's C++ code). As you say, sampling these integers across their full range would be relatively easy in C++, though passing them to R would require some care with the signed/unsigned transition. And I, at least, do want some method of getting the random seeds in my R session - see below.

I'm also surprised that `bit64` involves reinterpreting a `double` like that, I thought that would be implementation-specific behaviour. Maybe it's always the same on the platforms on which R compiles...

# Using R's RNG to define the seed

The clearest example is probably in _DropletUtils_ in the [`emptyDrops` function](https://github.com/MarioniLab/DropletUtils/blob/master/R/emptyDrops.R). This is single-cell genomics data where the C++ code loops across cells and does a series of Monte Carlo simulations to compute a per-cell p-value. Normally, I would just use R's PRNG within C++, i.e., initialize the PRNG at the start of the C++ function and run through all cells without resetting the PRNG or anything. 

The problem is that this code is potentially parallelized at the R-level. This is achieved via the the `bplapply` loop (from the _BiocParallel_ package), with a variety of parallelization schemes defined by the `BPPARAM` argument. This means that I get a different series of random numbers depending on the parallelization scheme, i.e., number of workers and how the jobs are partitioned across those workers. Obviously, this is not desirable as I would like to get the same results regardless of the chosen parallelization.

The solution is to seed a fresh PRNG for each cell. This guarantees that I get the same results regardless of how the jobs (one job per cell) are distributed across workers, as each cell's series of random numbers are unrelated to that of another cell's. Moreover, I'm using PCG32 with different streams, which provides an added level of reassurance with respect to statistical independence. The construction of the PRNG is done within the cell-by-cell loop in C++, but the generation of the seeds is still performed in the R code where (i) it can be affected by `set.seed()` and (ii) it is still serially executed and thus avoids weird changes in seeds in workers under some parallelization schemes, e.g., those using cluster job schedulers.

This approach does, however, rely on a lot of re-seeding, which means that statistical biases may manifest in the random numbers if I can only seed a part of the state space. Which brings us to #10 and this issue.

**Comment by LTLA** [comment]
If raw vectors are too unpalatable, I just did a refactor to use integer vectors - this is on the `integerseed` branch of my fork [here](https://github.com/LTLA/dqrng/tree/integerseed). As anticipated, some work was required to guarantee a sensible `uint32_t` to `int` conversion, which would otherwise be implementation-defined if we left it to a `static_cast`.

**Comment by LTLA** [comment]
**Edit, 5 hours later:** I spent some time thinking about the use of `seed_seq_fe` here - and I don't think it's necessary. From what I understand, the seed sequence is intended to "inflate" a small amount of entropy to get enough bits for seeding. But here, R's MT stream is already giving us decent random numbers for seeding. There's no need to complicate the _dqrng_ API with extra work and bug potential, e.g., signed vs. unsigned, how to guarantee randomness/full state space exploration when more than four integers are requested.

Perhaps the use of the seed sequence would be useful for handling the case where `dqset.seed()` is supplied with a single integer. Otherwise, the current changes in #10 (now switched to using integer vectors, as suggested) seem sufficient for all but the most advanced PRNG users.

**Comment by rstub** [comment]
Fixed in #10.

**Comment by LTLA** [comment]
For posterity's sake, I had a go at using _dqrng_ with multiple streams vs multiple seeds:

## streamer.cpp

```cpp
// [[Rcpp::plugins(cpp11)]]
#include <Rcpp.h>

// [[Rcpp::depends(dqrng)]]
#include "convert_seed.h"
#include "pcg_random.hpp"

// [[Rcpp::export(rng=false)]]
Rcpp::IntegerVector spew_raw_numbers(Rcpp::IntegerVector seed, int stream, int N=10000) {
    pcg32 rng(dqrng::convert_seed<uint64_t>(seed), stream);
    Rcpp::IntegerVector output(N);

    for (auto& o : output) {
        uint32_t sampled=rng();

        // Converting uint to int.
        constexpr uint32_t max_int=2147483647;
        if (sampled <= max_int) {
            o=static_cast<int>(sampled);
        } else {
            constexpr uint32_t max_uint=-1;
            o=-static_cast<int>(max_uint - sampled) - 1;
        }
    }

    return output;
}
```

## R code

```r
Rcpp::sourceCpp("streamer.cpp")

collected0 <- collected1 <- collected2 <- numeric(1000)
for (i in seq_along(collected1)) {
    out <- dqrng::generateSeedVectors(2)
        
    x1_1 <- spew_raw_numbers(out[[1]], 1, 20000)
    x1_1a <- head(x1_1, 10000)
    x1_1b <- tail(x1_1, 10000)

    x2_2 <- spew_raw_numbers(out[[2]], 2)
    x1_2 <- spew_raw_numbers(out[[1]], 2)

    # Switch functions to test different types of correlations:
#    FUN <- cor
#    FUN <- function(...) cor(..., method="spearman")
    FUN <- function(...) max(abs(ccf(..., plot=FALSE)$acf))

    collected0[i] <- FUN(x1_1a, x1_1b)
    collected1[i] <- FUN(x1_1a, x2_2)
    collected2[i] <- FUN(x1_1a, x1_2)
}

mean(abs(collected0))
mean(abs(collected1))
mean(abs(collected2))

max(abs(collected0))
max(abs(collected1))
max(abs(collected2))
```

Yeah, it's not very sophisticated, but it's the best I've got without further knowledge of the failure modes of this PRNGs. I really wish someone would chime in on imneme/pcg-cpp#49...

## TL; DR

For the amount of random numbers I'm generating, there is no observable differences in mean or maximum correlation from a jump-ahead approach on a single stream compared to multiple streams with the same seed or multiple streams with different seeds. So both our approaches are probably fine. Probably.
